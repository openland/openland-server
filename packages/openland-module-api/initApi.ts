import { Config } from 'openland-config/Config';
import * as bodyParser from 'body-parser';
import express from 'express';
import cors from 'cors';
// import morgan from 'morgan';
import * as Auth2 from '../openland-module-auth/authV2';
import * as Auth from '../openland-module-auth/providers/email';
import * as AuthGoogle from '../openland-module-auth/providers/google';
import { Schema } from './schema/Schema';
import { fetchWebSocketParameters, buildWebSocketContext } from './handlers/websocket';
import { errorHandler, QueryInfo } from '../openland-errors';
import { withAudit } from '../openland-module-auth/providers/email';
import { IDs } from './IDs';
import { Modules } from 'openland-modules/Modules';
import { initSafariPush } from '../openland-module-push/safari/handlers';
import { initAppHandlers } from '../openland-module-apps/Apps.handler';
import { ApolloServer } from 'apollo-server-express';
import { callContextMiddleware } from './handlers/context';
import * as http from 'http';
import { TokenChecker } from '../openland-module-auth/authV2';
import { parseCookies } from '../openland-utils/parseCookies';
import { decode } from 'openland-utils/base64';
import { AccessDeniedError } from 'openland-errors/AccessDeniedError';
import { createFuckApolloWSServer } from '../openland-mtproto3';
import * as url from 'url';
import { buildConcurrencyPool } from './buildConcurrencyPool';
import { withConcurrentcyPool } from 'openland-utils/ConcurrencyPool';
import { createNamedContext } from '@openland/context';
import { createLogger, withLogPath } from '@openland/log';
import { withReadOnlyTransaction, inTx } from '@openland/foundationdb';
import {
    GqlQueryIdNamespace,
    GqlTrace,
    gqlTraceNamespace,
    withGqlQueryId,
    withGqlTrace
} from '../openland-graphql/gqlTracer';
// import { AuthContext } from '../openland-module-auth/AuthContext';
import { uuid } from '../openland-utils/uuid';
// import { createMetric } from '../openland-module-monitoring/Metric';
// import { currentRunningTime } from '../openland-utils/timer';
import { withLifetime } from '@openland/lifetime';
// import { initIFTTT } from '../openland-module-ifttt/http.handlers';
import { InMemoryQueryCache } from '../openland-mtproto3/queryCache';
import { initZapier } from '../openland-module-zapier/http.handlers';
import { initVostokApiServer } from '../openland-mtproto3/vostok-api/vostokApiServer';
import { EventBus } from '../openland-module-pubsub/EventBus';
import { initOauth2 } from '../openland-module-oauth/http.handlers';
import { AuthContext } from '../openland-module-auth/AuthContext';
import { initPhoneAuthProvider } from '../openland-module-auth/providers/phone';
import { Shutdown } from '../openland-utils/Shutdown';
import { Metrics } from '../openland-module-monitoring/Metrics';
// import { Store } from '../openland-module-db/FDB';
// import { fetchNextDBSeq } from '../openland-utils/dbSeq';
// import { createFuckApolloWSServer } from '../openland-mtproto3';
// import { randomKey } from '../openland-utils/random';

// const loggerWs = createLogger('ws');
const integrationCtx = createNamedContext('integration-ctx');
const logger = createLogger('api-module');
// const authMetric = createMetric('auth-metric', 'average');
// const authMetricCtx = createNamedContext('ws');

// const onGqlQuery = createHyperlogger<{ operationName: string, duration: number }>('gql_query_tracing');

export async function initApi(isTest: boolean) {
    const rootCtx = createNamedContext('init');
    logger.log(rootCtx, 'Starting...');

    //
    // Fetching Port
    //
    let port = process.env.PORT;
    let dport = 9000;
    if (port !== undefined && port !== '') {
        dport = parseInt(process.env.PORT as string, 10);
    }
    if (isTest) {
        dport = 0;
    }

    //
    // Middlewares
    //

    const app = express();

    app.enable('trust proxy');

    // Basic Configuration
    if (!isTest) {
        app.use(cors());
        // app.use(morgan('tiny'));
        // app.use(compression());
    }

    // To avoid logging on this route
    app.get('/', (req, res) => res.status(200).send('Welcome to Openland API!'));
    app.get('/status', (req, res) => res.status(200).send('Welcome to Openland API!'));
    app.get('/favicon.ico', (req, res) => res.sendStatus(404));
    app.get('/robots.txt', (req, res) => res.sendStatus(404));

    //
    // Authenticaton
    //
    app.post('/v2/auth', Auth2.JWTChecker, bodyParser.json(), Auth2.Authenticator);

    app.post('/auth/sendCode', bodyParser.json(), withAudit(Auth.sendCode));
    app.post('/auth/checkCode', bodyParser.json(), withAudit(Auth.checkCode));
    app.post('/auth/getAccessToken', bodyParser.json(), withAudit(Auth.getAccessToken));
    app.post('/auth/google/getAccessToken', bodyParser.json(), withAudit(AuthGoogle.getAccessToken));

    initPhoneAuthProvider(app);

    //
    //  Safari push
    //
    initSafariPush(app);

    //
    //  Apps
    //
    initAppHandlers(app);

    // IFTTT api
    // initIFTTT(app);

    // Zapier api
    initZapier(app);

    // Ouath2 api
    initOauth2(app);

    //
    // Semaphore bot
    //
    app.post('/semaphorebot', bodyParser.json(), (async (req, res) => {
        await inTx(integrationCtx, async (ctx) => {
            let chatId = IDs.Conversation.parse('zoqLwdzr6VHelWZOR6JatW4aak');
            let botId = IDs.User.parse('vmM5pE5lQ4udYLr6AOLBhdQDJK');
            let data = req.body;

            if (data.result === 'passed') {
                let text = `${data.commit.author_name} ${data.event === 'deploy' ? 'deployed' : 'build'} :tada: - ${data.commit.message} to ${data.project_name}`;

                await Modules.Messaging.sendMessage(ctx, chatId, botId, {message: text});
            }
        });
    }));

    //
    // Graphana alerts
    //
    app.post('/graphanaAlerts', bodyParser.json(), (async (req, res) => {
        let authRaw = req.headers.authorization;
        if (!authRaw || decode(authRaw!.split(' ')[1]) !== 'grfana:d138df93-758a-4e8c-be8f-e46ecf09eeb4') {
            throw new AccessDeniedError();
        }
        await inTx(integrationCtx, async (ctx) => {
            let chatId = IDs.Conversation.parse('M6Pl7R30rECQn7a9OP4MHrqYdo');
            let botId = IDs.User.parse('LOaDEWDjZQfjVm3P7Ro4CYgMAD');
            let data = req.body;

            let text = data.title + '\n' + data.message + (data.imageUrl ? '\n' + data.imageUrl : '') + (data.state === 'ok' ? '\nОтпустило кажется' : '');
            await Modules.Messaging.sendMessage(ctx, chatId, botId, {message: text});
        });
    }));

    const Server = new ApolloServer({
        schema: Schema(),
        introspection: true,
        tracing: Config.environment !== 'production',
        formatError: (err: any) => {
            logger.warn(rootCtx, err);
            return {
                ...errorHandler(err),
                locations: err.locations,
                path: err.path
            };
        },
        context: async (context) => {
            let ctx = context as any;
            // WS
            if (ctx.connection) {
                let wsctx = ctx.connection.context;
                let ctx2 = buildWebSocketContext(
                    wsctx || {},
                    context.req.header('X-Forwarded-For'),
                    context.req.header('X-Client-Geo-LatLong'),
                    context.req.header('X-Client-Geo-Location')
                );

                return withConcurrentcyPool(ctx2, buildConcurrencyPool(ctx2));
            }
            await TokenChecker(context.req, context.res);
            await callContextMiddleware(isTest, context.req, context.res);

            return context.res.locals.ctx;
        },
        // subscriptions: {
        //     onConnect: async (connectionParams, websocket, context) => {
        //         return await fetchWebSocketParameters(connectionParams, websocket);
        //     },
        //     path: '/api',
        // },
        playground: {
            endpoint: process.env.NODE_ENV === 'production' ? '' : 'http://localhost:3000/graphql',
            settings: {
                'request.credentials': 'include'
            } as any
        }
    });

    let server: http.Server;
    // Starting Api
    if (dport > 0) {
        logger.log(rootCtx, 'Binding to port ' + dport);

        let formatError = (err: any, info?: QueryInfo) => {
            logger.warn(rootCtx, 'api_error', err, info);
            return {
                ...errorHandler(err, info),
                locations: err.locations,
                path: err.path
            };
        };

        // Starting server
        const httpServer = http.createServer(app);

        Server.applyMiddleware({app, path: '/graphql'});
        Server.applyMiddleware({app, path: '/api'});

        async function saveTrace(trace: GqlTrace) {
            if (trace.duration >= 1500) {
                // await inTx(rootCtx, async _ctx => {
                //     let id = await fetchNextDBSeq(_ctx, 'gql-trace');
                //     await Store.GqlTrace.create(_ctx, id, { traceData: trace });
                // });
            }
        }
        // const wsCtx = createNamedContext('ws-gql');
        let fuckApolloWS = await createFuckApolloWSServer({
            server: undefined, // httpServer ,
            path: '/api',
            executableSchema: Schema(),
            queryCache: new InMemoryQueryCache(),
            onAuth: async (params, req) => {
                // const start = currentRunningTime();
                try {
                    if (!params || Object.keys(params).length === 0 && req.headers.cookie && req.headers.cookie.length > 0) {
                        let cookies = parseCookies(req.headers.cookie || '');
                        return await fetchWebSocketParameters({'x-openland-token': cookies['x-openland-token']}, null);
                    }
                    return await fetchWebSocketParameters(params, null);
                } finally {
                    // const delta = currentRunningTime() - start;
                    // authMetric.add(authMetricCtx, delta);
                }
            },
            context: async (params, operation, req) => {
                let opId = uuid();
                let ctx = buildWebSocketContext(
                    params || {},
                    req.headers['x-forwarded-for'] as string,
                    req.headers['x-client-geo-latlong'] as string,
                    req.headers['x-client-geo-location'] as string
                );
                ctx = withReadOnlyTransaction(ctx);
                ctx = withLogPath(ctx, `query ${opId} ${operation.operationName || ''}`);
                ctx = withGqlQueryId(ctx, opId);
                ctx = withGqlTrace(ctx, `query ${opId} ${operation.operationName || ''}`);
                return withConcurrentcyPool(ctx, buildConcurrencyPool(ctx));
            },
            subscriptionContext: async (params, operation, firstCtx, req) => {
                let opId = firstCtx ? GqlQueryIdNamespace.get(firstCtx)! : uuid();
                let ctx = buildWebSocketContext(
                    params || {},
                    req.headers['x-forwarded-for'] as string,
                    req.headers['x-client-geo-latlong'] as string,
                    req.headers['x-client-geo-location'] as string,
                );
                ctx = withReadOnlyTransaction(ctx);
                ctx = withLogPath(ctx, `subscription ${opId} ${operation.operationName || ''}`);
                ctx = withGqlQueryId(ctx, opId);
                ctx = withGqlTrace(ctx, `subscription ${opId} ${operation.operationName || ''}`);
                ctx = withLifetime(ctx);
                return withConcurrentcyPool(ctx, buildConcurrencyPool(ctx));
            },
            onOperation: async (ctx, operation) => {
                // if (!isTest) {
                //     let opId = GqlQueryIdNamespace.get(ctx) || 'unknown query';
                //     if (AuthContext.get(ctx).uid) {
                //         logger.log(wsCtx, `GraphQL ${opId} [#${AuthContext.get(ctx).uid}]: ${JSON.stringify(operation)}`);
                //     } else {
                //         logger.log(wsCtx, `GraphQL ${opId} [#ANON]: ${JSON.stringify(operation)}`);
                //     }
                // }
                Metrics.GQLRequests.inc();
            },
            onOperationFinish: (ctx, operation, duration) => {
                // let trace = gqlTraceNamespace.get(ctx);
                // if (trace) {
                //     trace.onRequestFinish();
                //     await saveTrace(trace.getTrace());
                // }
                Metrics.GQLRequests.dec();
                Metrics.GQLRequestTime.add(duration, uuid(), 10000);
            },
            onEventResolveFinish: async (ctx, operation, duration) => {
                let trace = gqlTraceNamespace.get(ctx);
                if (trace) {
                    trace.onRequestFinish();
                    await saveTrace(trace.getTrace());
                }
                if (operation.operationName) {
                    Metrics.GQLRequestTime.add(duration, uuid(), 10000);
                }
            },
            formatResponse: (value, operation, ctx) => {
                let auth = AuthContext.get(ctx);
                let uid = auth.uid;
                let oid = auth.oid;

                let queryInfo: QueryInfo = {
                    uid,
                    oid,
                    transport: 'ws',
                    query: JSON.stringify(operation)
                };

                let errors: any[] | undefined;
                if (value.errors) {
                    errors = value.errors && value.errors.map((e: any) => formatError(e, queryInfo));
                }
                return ({
                    ...value,
                    errors: errors,
                });
            }
        });
        let vostok = initVostokApiServer({
            server: undefined, // httpServer ,
            path: '/api',
            executableSchema: Schema(),
            queryCache: new InMemoryQueryCache(),
            onAuth: async (token) => {
                return await fetchWebSocketParameters({'x-openland-token': token}, null);
            },
            context: async (params, operation) => {
                let opId = uuid();
                let ctx = buildWebSocketContext(params || {});
                ctx = withReadOnlyTransaction(ctx);
                ctx = withLogPath(ctx, `query ${opId} ${operation.operationName || ''}`);
                ctx = withGqlQueryId(ctx, opId);
                ctx = withGqlTrace(ctx, `query ${opId} ${operation.operationName || ''}`);
                return ctx;
            },
            subscriptionContext: async (params, operation, firstCtx) => {
                let opId = firstCtx ? GqlQueryIdNamespace.get(firstCtx)! : uuid();
                let ctx = buildWebSocketContext(params || {});
                ctx = withReadOnlyTransaction(ctx);
                ctx = withLogPath(ctx, `subscription ${opId} ${operation.operationName || ''}`);
                ctx = withGqlQueryId(ctx, opId);
                ctx = withGqlTrace(ctx, `subscription ${opId} ${operation.operationName || ''}`);
                ctx = withLifetime(ctx);
                return ctx;
            },
            onOperation: async (ctx, operation) => {
                // noop
            },
            formatResponse: async (value, operation, ctx) => {
                let auth = AuthContext.get(ctx);
                let uid = auth.uid;
                let oid = auth.oid;

                let queryInfo: QueryInfo = {
                    uid,
                    oid,
                    transport: 'ws',
                    query: JSON.stringify(operation)
                };

                let errors: any[] | undefined;
                if (value.errors) {
                    errors = value.errors && value.errors.map((e: any) => formatError(e, queryInfo));
                }
                return ({
                    ...value,
                    errors: errors,
                });
            }
        });

        EventBus.subscribe('auth_token_revoke', (data: { tokens: { uuid: string, salt: string }[] }) => {
            for (let token of data.tokens) {
                for (let entry of fuckApolloWS.sessions.entries()) {
                    let [, session] = entry;
                    if (session.authParams && session.authParams.tid && session.authParams.tid === token.uuid) {
                        session.close();
                        session.stopAllOperations();
                    }
                }

                for (let entry of vostok.sessions.sessions.entries()) {
                    let [, session] = entry;
                    if (session.authParams && session.authParams.tid && session.authParams.tid === token.uuid) {
                        session.close();
                    }
                }
            }
        });

        httpServer.on('upgrade', (request, socket, head) => {
            const pathname = url.parse(request.url).pathname;

            if (pathname === '/api') {
                fuckApolloWS.ws.handleUpgrade(request, socket, head, (_ws) => {
                    fuckApolloWS.ws.emit('connection', _ws, request);
                });
            } else if (pathname === '/gql_ws') {
                fuckApolloWS.ws.handleUpgrade(request, socket, head, (_ws) => {
                    fuckApolloWS.ws.emit('connection', _ws, request);
                });
            } else if (pathname === '/vostok') {
                vostok.ws.handleUpgrade(request, socket, head, (_ws) => {
                    vostok.ws.emit('connection', _ws, request);
                });
            } else {
                socket.destroy();
            }
        });
        httpServer.listen(dport);
        Shutdown.registerWork({
            name: 'http-server',
            shutdown: async () => {
                httpServer.close();
            }
        });
        server = httpServer;
    } else {
        server = await new Promise((resolver) => app.listen(0, () => resolver()));
    }

    return server;
}
