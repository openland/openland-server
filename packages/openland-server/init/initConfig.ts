import fs from 'fs';
import WebPush from 'web-push';

class AppConfig {
    webPush: { private: string, public: string } | undefined;
    apple: { teamId: string, key: string, keyId: string, bundles: string[] }[] | undefined;
    google: { privateKey: string, projectId: string, clientEmail: string, databaseURL: string, packages: string[] }[] | undefined;
}

export const AppConfiuguration = new AppConfig();

interface CertsConfig {
    'web-push': {
        private: string,
        public: string;
    } | undefined | null;
    'apple': {
        teams: {
            key: string;
            'key-id': string;
            'team-id': string;
            bundles: string[];
        }[] | undefined | null
    } | undefined | null;
    'google': {
        accounts: {
            key: {
                type: string
                project_id: string
                private_key_id: string
                private_key: string
                client_email: string
                client_id: string
                auth_uri: string
                token_uri: string
                auth_provider_x509_cert_url: string
                client_x509_cert_url: string
            },
            endpoint: string,
            packages: string[]
        }[] | undefined | null;
    } | undefined | null;
}

if (process.env.PUSH_CERTS_PATH) {
    let config = JSON.parse(fs.readFileSync(process.env.PUSH_CERTS_PATH, 'utf-8')) as CertsConfig;

    // Loading Web Push
    if (config['web-push']) {
        AppConfiuguration.webPush = { private: config['web-push']!!.private, public: config['web-push']!!.public };
        WebPush.setVapidDetails(
            'mailto:support@openland.com',
            AppConfiuguration.webPush.public,
            AppConfiuguration.webPush.private
        );
    }
    if (config.apple && config.apple.teams && config.apple.teams.length > 0) {
        let t = config.apple!!.teams!!;
        AppConfiuguration.apple = t.map((v) => ({ teamId: v['team-id'], keyId: v['key-id'], key: v.key, bundles: v.bundles }));
    }

    if (config.google && config.google.accounts) {
        AppConfiuguration.google = config.google.accounts.map(a => ({ privateKey: a.key.private_key, projectId: a.key.project_id, clientEmail: a.key.client_email, databaseURL: a.endpoint, packages: a.packages }));
    }
}