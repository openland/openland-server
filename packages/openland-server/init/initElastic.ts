import { enableIndexer } from '../indexing';
export async function initElastic() {
    if (process.env.ELASTIC_ENABLE_INDEXING !== 'false') {
        await enableIndexer();
    }
}