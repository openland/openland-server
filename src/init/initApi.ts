import * as bodyParser from 'body-parser';
import express from 'express';
import cors from 'cors';
import morgan from 'morgan';
import compression from 'compression';
import * as Auth2 from '../handlers/authV2';
import { schemaHandler } from '../handlers/schema';
import { SubscriptionServer } from 'subscriptions-transport-ws';
import { Schema } from '../api';
import { execute, subscribe } from 'graphql';
import { fetchWebSocketParameters, buildWebSocketContext } from '../handlers/websocket';
import { errorHandler } from '../errors';
import { Rate } from '../utils/rateLimit';
import { Server as HttpServer } from 'http';
import { ApolloEngine } from 'apollo-engine';
import { delay } from '../utils/timer';

export async function initApi(isTest: boolean) {

    console.info('Starting...');

    //
    // Fetching Port
    //
    let port = process.env.PORT;
    let dport = 9000;
    if (port !== undefined && port !== '') {
        dport = parseInt(process.env.PORT as string, 10);
    }
    if (isTest) {
        dport = 0;
    }

    // Fetchin Apollo Engine
    let engineKey = process.env.APOLLO_ENGINE_KEY;
    if (engineKey) {
        console.log('Starting with Apollo Engine');
    }

    //
    // Middlewares
    //

    const app = express();

    app.enable('trust proxy');

    // To avoid logging on this route
    app.get('/', (req, res) => res.send('Welcome to Openland API!'));
    app.get('/status', (req, res) => res.send('Welcome to Openland API!'));
    app.get('/healthz', (req, res) => res.send('Welcome to Openland API!'));
    app.get('/favicon.ico', (req, res) => res.send(404));

    // Basic Configuration
    if (!isTest) {
        app.use(cors());
        app.use(morgan('tiny'));
        app.use(compression());
    }

    //
    // API
    //
    let graphqlMiddleware = schemaHandler(isTest, !!engineKey);
    app.use('/api', Auth2.TokenChecker, bodyParser.json({ limit: '5mb' }), graphqlMiddleware);
    app.use('/graphql', Auth2.TokenChecker, bodyParser.json({ limit: '5mb' }), graphqlMiddleware);

    //
    // Authenticaton
    //
    app.post('/v2/auth', Auth2.JWTChecker, bodyParser.json(), Auth2.Authenticator);

    // Starting Api
    if (dport > 0) {
        console.info('Binding to port ' + dport);

        let formatError = (err: any) => {
            console.warn(err);
            return {
                ...errorHandler(err),
                locations: err.locations,
                path: err.path
            };
        };

        function createWebSocketServer(server: HttpServer) {
            new SubscriptionServer({
                schema: Schema,
                execute,
                subscribe,
                keepAlive: 10000,
                onConnect: async (args: any, webSocket: any) => {
                    webSocket.__params = await fetchWebSocketParameters(args, webSocket);
                },
                onOperation: async (message: any, params: any, webSocket: any) => {
                    if (!isTest) {
                        if (webSocket.__params.uid) {
                            console.log('WS GraphQL [#' + webSocket.__params.uid + ']: ' + JSON.stringify(message.payload));
                        } else {
                            console.log('WS GraphQL [#ANON]: ' + JSON.stringify(message.payload));
                        }
                    }

                    let clientId = '';

                    if (webSocket.__params.uid) {
                        clientId = 'user_' + webSocket.__params.uid;
                    } else {
                        clientId = 'ip_' + webSocket._socket.remoteAddress;
                    }

                    let handleStatus = Rate.WS.canHandle(clientId);

                    if (!handleStatus.canHandle) {
                        if (handleStatus.delay) {
                            Rate.WS.hit(clientId);
                            await delay(handleStatus.delay);
                        } else {
                            throw new Error('Rate limit!');
                        }
                    } else {
                        Rate.HTTP.hit(clientId);
                    }

                    return {
                        ...params,
                        context: buildWebSocketContext(webSocket.__params),
                        formatResponse: (value: any) => ({
                            ...value,
                            errors: value.errors && value.errors.map(formatError),
                        })
                    };
                }
            }, { server: server, path: '/api' });
        }

        if (engineKey) {
            // Starting server with Apollo Engine
            let server = new HttpServer(app);
            const engine = new ApolloEngine({
                apiKey: engineKey
            });
            engine.listen({
                port: dport,
                httpServer: server,
                graphqlPaths: ['/graphql', '/api']
            }, () => { createWebSocketServer(server); });
        } else {
            // Starting server
            createWebSocketServer(app.listen(dport));
        }

    } else {
        await new Promise((resolver) => app.listen(0, () => resolver()));
    }

    return app;
}
