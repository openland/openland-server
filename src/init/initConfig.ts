import fs from 'fs';
import WebPush from 'web-push';

class AppConfig {
    webPush: { private: string, public: string } | undefined;
}

export const AppConfiuguration = new AppConfig();

interface CertsConfig {
    'web-push': {
        private: string,
        public: string;
    } | undefined | null;
}

if (process.env.PUSH_CERTS_PATH) {
    let config = JSON.parse(fs.readFileSync(process.env.PUSH_CERTS_PATH, 'utf-8')) as CertsConfig;
    
    // Loading Web Push
    if (config['web-push']) {
        AppConfiuguration.webPush = { private: config['web-push']!!.private, public: config['web-push']!!.public };
        WebPush.setVapidDetails(
            'mailto:support@openland.com',
            AppConfiuguration.webPush.public,
            AppConfiuguration.webPush.private
        );
    }
}