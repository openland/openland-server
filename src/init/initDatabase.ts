import * as fs from 'fs';
import * as cp from 'child_process';
import * as db from '../connector';
import { redisClient } from '../modules/redis/redis';

export async function initDatabase(isTest: boolean, drop: boolean) {
    if ((isTest && drop) || (process.env.RECREATE_DB === 'true' && process.env.NODE_ENV === 'development')) {
        // Dropping Database
        await db.connection.getQueryInterface().dropAllTables();
        await db.connection.getQueryInterface().dropAllSchemas();

        // Dropping Redis
        if (redisClient()) {
            await redisClient()!!.flushall();
        }
    }
    if (process.env.NODE_ENV === 'development') {
        if (!isTest) { console.info('Connecting to database in DEVELOPMENT mode'); }

        if (process.env.RECREATE_DB === 'true') {
            // TODO: Dropping elastic search

            if (fs.existsSync('./dumps/dump.sql')) {
                console.warn('Recreating database');
                // cp.execSync('psql -q -h localhost -U kor_ka -d postgres -f ./dumps/dump.sql', { stdio: 'inherit' });
                cp.execSync('psql -q -h localhost -U ' + process.env.DATABASE_USER + ' -d postgres -f ./dumps/dump.sql', { stdio: 'inherit' });
                console.warn('Database imported');
            } else {
                throw new Error('Unable to find ./dumps/dump.sql');
            }

            // Resetting locks and readers
            await db.connection.query('TRUNCATE TABLE locks;', { logging: !isTest });
            await db.connection.query('TRUNCATE TABLE reader_states;', { logging: !isTest });
        }
    } else {
        if (!isTest) { console.info('Connecting to database in RELEASE mode'); }
    }
    await db.migrate();
}