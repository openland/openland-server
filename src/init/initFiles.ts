import { checkFilesConfig } from '../modules/files';

export async function initFiles() {
    await checkFilesConfig();
}