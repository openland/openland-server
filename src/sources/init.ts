import * as fdb from 'foundationdb';
import * as fs from 'fs';

fdb.setAPIVersion(510);

let db: fdb.Database;
if (process.env.FOUNDATION_DB) {
    fs.writeFileSync('foundation.clusterfile', process.env.FOUNDATION_DB);
    db = fdb.openSync('foundation.clusterfile');
} else {
    db = fdb.openSync();
}

export const FDBConnection =
    db.withKeyEncoding(fdb.encoders.tuple)
        .withValueEncoding(fdb.encoders.json);