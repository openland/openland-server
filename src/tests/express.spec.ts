import prepareApp from './utils/prepareApp';
import { initApi } from '../init/initApi';

beforeAll(async () => {
    await prepareApp();
    await initApi(true);
});

describe('API Server', () => {
    it('should authenticate correctly', () => {
        // TODO
    });
});