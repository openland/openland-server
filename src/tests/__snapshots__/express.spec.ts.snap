// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Server should handle graphql queries 1`] = `
Object {
  "data": Object {
    "me": null,
  },
}
`;

exports[`API Server should handle graphql queries 2`] = `
Object {
  "data": Object {
    "me": null,
  },
}
`;

exports[`API Server should handle graphql queries with tokens 1`] = `
Object {
  "data": Object {
    "me": Object {
      "id": "GvNrAkJ7rbFGQmJJJJdyhKaBr4",
    },
  },
}
`;

exports[`API Server should handle graphql queries with tokens in cookies 1`] = `
Object {
  "data": Object {
    "me": Object {
      "id": "GvNrAkJ7rbFGQmJJJJdyhKaBr4",
    },
  },
}
`;
